{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lesson 35: More pytest and continuous integration\n",
    "\n",
    "(c) 2017 Justin Bois and Davi Ortega. This work is licensed under a [Creative Commons Attribution License CC-BY 4.0](https://creativecommons.org/licenses/by/4.0/). All code contained herein is licensed under an [MIT license](https://opensource.org/licenses/MIT).\n",
    "\n",
    "*This tutorial was generated from a Jupyter notebook.  You can download the notebook [here](l35_pytest_and_CI.ipynb).*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Handling odd behaviors\n",
    "\n",
    "To explore another feature of `pytest`, we'll consider another aspect of our `n_neg()` function.  Specifically, what should we do if an invalid sequence is entered?\n",
    "\n",
    "A sensible thing to do in this case is to make our software *throw* a `RuntimeError`.  \n",
    "\n",
    "Again, in designing our test, we need to think about what constitutes an invalid sequence.  We'll only allow the 20 symbols for the residues that we used in [previous lessons](l15_exceptions_and_error_handling.html#Checking-input) and present in the `bioinfo_dicts.py` module.  So, we adjust our test function accordingly.  We cannot use the **`assert`** statement to check for proper error handling, so we use the `pytest.raises()` function.  This function takes as its first argument the type of exception expected, and a string containing the code to be run to give the exception."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A note on assertions vs raising exceptions\n",
    "\n",
    "It is important to draw the distinction between assertions and raising exceptions in your code.  \n",
    "* You should raise **exceptions** when you are checking inputs to your function.  I.e., you are checking to make sure the user is using the function properly.\n",
    "* You should use **assertions** to make sure the function operates as expected for given input.\n",
    "\n",
    "We should then alter the code of the `test_seq_feature.py` to include our expectation that the program should throw a `RuntimeError` if an invalid sequence is entered:\n",
    "\n",
    "```python\n",
    "def test_n_neg_for_invalid_amino_acid():\n",
    "    with pytest.raises(RuntimeError) as excinfo:\n",
    "        seq_features.n_neg('Z')\n",
    "    excinfo.match(\"Z is not a valid amino acid\")\n",
    "```\n",
    "\n",
    "We also have to include `import pytest` at the beginning of the `test_seq_feature.py` file. It is clear that if `Z` is passed as the input sequence, the program should throw a `RuntimeError` saying: *\"Z is an invalid sequence\"*. Let's test:\n",
    "\n",
    "```\n",
    "$ pytest -v\n",
    "=============================== test session starts ===============================\n",
    "platform darwin -- Python 3.6.1, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /Users/Justin/anaconda/bin/python\n",
    "cachedir: .cache\n",
    "rootdir: /Users/Justin/git/programming_bootcamp/2017/lessons, inifile:\n",
    "collected 5 items \n",
    "\n",
    "test_seq_features.py::test_n_neg_for_single_E_or_D PASSED\n",
    "test_seq_features.py::test_n_neg_for_empty_sequence PASSED\n",
    "test_seq_features.py::test_n_neg_for_longer_sequences PASSED\n",
    "test_seq_features.py::test_n_neg_for_lower_case_sequences PASSED\n",
    "test_seq_features.py::test_n_neg_for_invalid_amino_acid FAILED\n",
    "\n",
    "==================================== FAILURES =====================================\n",
    "________________________ test_n_neg_for_invalid_amino_acid ________________________\n",
    "\n",
    "    def test_n_neg_for_invalid_amino_acid():\n",
    "        with pytest.raises(RuntimeError) as excinfo:\n",
    ">           seq_features.n_neg('Z')\n",
    "E           Failed: DID NOT RAISE <class 'RuntimeError'>\n",
    "\n",
    "test_seq_features.py:24: Failed\n",
    "======================= 1 failed, 4 passed in 0.06 seconds ========================\n",
    "```\n",
    "\n",
    "Despite that all other four tests still pass, the last one fails because our program does not know yet how to throw a `RuntimeError` when it receives an invalid sequence as input. Let's fix that:\n",
    "\n",
    "```python\n",
    "def n_neg(seq):\n",
    "    \"\"\"Number of negative residues a protein sequence\"\"\"\n",
    "\n",
    "    # Convert sequence to upper case\n",
    "    seq = seq.upper()\n",
    "\n",
    "    if seq == 'Z':\n",
    "        raise RuntimeError('Z is not a valid amino acid.')\n",
    "\n",
    "    # Count E's and D's, since these are the negative residues\n",
    "    return seq.count('E') + seq.count('D')\n",
    "```\n",
    "\n",
    "re-running the test:\n",
    "\n",
    "```\n",
    "$ pytest -v\n",
    "=============================== test session starts ===============================\n",
    "platform darwin -- Python 3.6.1, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /Users/Justin/anaconda/bin/python\n",
    "cachedir: .cache\n",
    "rootdir: /Users/Justin/git/programming_bootcamp/2017/lessons, inifile:\n",
    "collected 5 items \n",
    "\n",
    "test_seq_features.py::test_n_neg_for_single_E_or_D PASSED\n",
    "test_seq_features.py::test_n_neg_for_empty_sequence PASSED\n",
    "test_seq_features.py::test_n_neg_for_longer_sequences PASSED\n",
    "test_seq_features.py::test_n_neg_for_lower_case_sequences PASSED\n",
    "test_seq_features.py::test_n_neg_for_invalid_amino_acid PASSED\n",
    "\n",
    "============================ 5 passed in 0.03 seconds =============================\n",
    "```\n",
    "\n",
    "Obviously, this is not a very robust fix; it only works if the invalid amino acid is `Z`. We need a smarter way to fix this. What about using the bioinformatics dictionary from before?\n",
    "\n",
    "```python\n",
    "import bioinfo_dicts\n",
    "\n",
    "def n_neg(seq):\n",
    "    \"\"\"Number of negative residues a protein sequence\"\"\"\n",
    "\n",
    "    # Convert sequence to upper case\n",
    "    seq = seq.upper()\n",
    "\n",
    "    # Check for a valid sequence\n",
    "    for aa in seq:\n",
    "        if aa not in bioinfo_dicts.aa.keys():\n",
    "            raise RuntimeError(aa + ' is not a valid amino acid.')\n",
    "\n",
    "    # Count E's and D's, since these are the negative residues\n",
    "    return seq.count('E') + seq.count('D')\n",
    "```\n",
    "\n",
    "running `pytest` one more time:\n",
    "\n",
    "```\n",
    "$  pytest -v\n",
    "============================ 5 passed in 0.03 seconds =============================\n",
    "Justin@JBMacBook [~/git/programming_bootcamp/2017/lessons]\n",
    "% py.test -v                                                                 (618)\n",
    "=============================== test session starts ===============================\n",
    "platform darwin -- Python 3.6.1, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /Users/Justin/anaconda/bin/python\n",
    "cachedir: .cache\n",
    "rootdir: /Users/Justin/git/programming_bootcamp/2017/lessons, inifile:\n",
    "collected 5 items \n",
    "\n",
    "test_seq_features.py::test_n_neg_for_single_E_or_D PASSED\n",
    "test_seq_features.py::test_n_neg_for_empty_sequence PASSED\n",
    "test_seq_features.py::test_n_neg_for_longer_sequences PASSED\n",
    "test_seq_features.py::test_n_neg_for_lower_case_sequences PASSED\n",
    "test_seq_features.py::test_n_neg_for_invalid_amino_acid PASSED\n",
    "\n",
    "============================ 5 passed in 0.03 seconds =============================\n",
    "```\n",
    "\n",
    "Hurray! Everything passed beautifully."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary of TDD\n",
    "\n",
    "Now that you have some experience with TDD and have an idea about what it is and how it works, let's formalize things by writing out the basic principles of test-driven development.\n",
    "\n",
    "1. Build your software out of **small functions** that do **one specific thing**.\n",
    "2. Build unit tests for all of your functions.\n",
    "3. Whenever you want to make any enhancements of adjustments to your code, write tests for it **first**.\n",
    "4. Whenever you encounter a bug, write tests for it that reproduce the behavior and then fix the code to make the entire test suite to pass."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Continuous Integration.\n",
    "\n",
    "Fine. Now you can write tests and functions that can pass those tests and push all that to a git repository so other people can use it too.\n",
    "\n",
    "However, one of the most important forces behind software development is collaboration. Chances are that even in small projects you will collaborate with other people during development or that another researcher will suggest changes in the future.\n",
    "\n",
    "This is where **continuous integration(CI)** comes to play.\n",
    "\n",
    "As you will probably reuse thousands of lines of code written by someone else in your program (have you stopped to think how many lines of code are invoked when you write `import numpy`?), others will also benefit from the lines of code you are writing right now. To make sure your code is up to snuff, the software community developed the concepts around unit testing and to facilitate collaborative work they developed git. **CI** is the combination of the two... continuously.\n",
    "\n",
    "Ideally every time someone submits changes to the remote repository (`git push`) somebody else checks the code by first running all the tests again and making sure it does pass them all. CI is the automation of this ideal.\n",
    "\n",
    "Here we will use one of the services available in GitHub for CI called **Travis CI**. Since **CI** is to simplify our lives, the folks from Travis and GitHub made the integration as easy as possible. So let's put the program that we just wrote in CI."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### New Travis CI account.\n",
    "\n",
    "First, go to the [travis-ci.org](http://travis-ci.org/) and sign up with your Github account.\n",
    "\n",
    "<img src='l35_travis_sign_up.png' width=\"800px\">\n",
    "\n",
    "Then, authorize the Travis app:\n",
    "\n",
    "<img src='l35_auth_travis.png' width=\"400px\">\n",
    "\n",
    "and then you now have your own page at travis-ci.org. Great. But it looks fairly empty there. We will come back to this."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Recap on how to make a new repository on Github\n",
    "\n",
    "Because we're now going through the process of setting up CI for a repository, let's recap how to set up a new repository hosted at GitHub. Chances are that your working directory is full of other scripts and whatnot. Let's start a new one for the sake of organization.\n",
    "\n",
    "Go to your Github page and click in `new repository` and give it a name. For this demo, we will use `protfeature` as our name.  \n",
    "\n",
    "<img src='l35_name_your_repo.png' width=\"700px\">\n",
    "\n",
    "and let's initialize the repo with the default specs:\n",
    "\n",
    "<img src='l35_init_repo_specs.png' width=\"700px\">\n",
    "\n",
    "Now, let's clone this repository into our own machine with `$ git clone` and copy the programs `seq_features.py`, `test_seq_features.py` and `bioinfo_dicts.py` to the directory created during cloning. If you followed our suggestion the name of the directory is: `protfeature`. Your directory should look like this:\n",
    "\n",
    "```\n",
    "-rw-r--r--+ 1 Justin  staff   1.0K Jun 15 22:38 LICENSE\n",
    "-rw-r--r--+ 1 Justin  staff    13B Jun 15 22:38 README.md\n",
    "-rw-r--r--+ 1 Justin  staff   1.0K Jun 15 22:36 bioinfo_dicts.py\n",
    "-rw-r--r--+ 1 Justin  staff   428B Jun 15 22:36 seq_features.py\n",
    "-rw-r--r--+ 1 Justin  staff   669B Jun 15 22:36 test_seq_features.py\n",
    "```\n",
    "\n",
    "before we proceed let's make sure this is working. `pytest` it!!!\n",
    "\n",
    "```\n",
    "$ pytest\n",
    "=============================== test session starts ===============================\n",
    "platform darwin -- Python 3.6.1, pytest-3.0.7, py-1.4.33, pluggy-0.4.0\n",
    "rootdir: /Users/Justin/git/protfeature, inifile:\n",
    "collected 5 items \n",
    "\n",
    "test_seq_features.py .....\n",
    "\n",
    "============================ 5 passed in 0.02 seconds =============================\n",
    "```\n",
    "\n",
    "Great! It works but you probably noticed that it generates a bunch of other files:\n",
    "\n",
    "```\n",
    "$ ls -l\n",
    "total 56\n",
    "-rw-r--r--  1 ortegad  staff  1068 Jun 14 12:49 LICENSE\n",
    "-rw-r--r--  1 ortegad  staff    35 Jun 14 12:49 README.md\n",
    "drwxr-xr-x  3 ortegad  staff   102 Jun 14 12:51 __pycache__\n",
    "-rw-r--r--  1 ortegad  staff  1005 Jun 14 12:51 bioinfo_dicts.py\n",
    "-rw-r--r--  1 ortegad  staff  1033 Jun 14 12:51 bioinfo_dicts.pyc\n",
    "-rw-r--r--  1 ortegad  staff   423 Jun 14 12:51 seq_features.py\n",
    "-rw-r--r--  1 ortegad  staff   601 Jun 14 12:51 seq_features.pyc\n",
    "-rw-r--r--  1 ortegad  staff   841 Jun 14 12:51 test_seq_features.py\n",
    "```\n",
    "\n",
    "If you do a `git status` all these files will be there waiting to be staged:\n",
    "\n",
    "```\n",
    "$ git status\n",
    "On branch master\n",
    "Your branch is up-to-date with 'origin/master'.\n",
    "Untracked files:\n",
    "  (use \"git add <file>...\" to include in what will be committed)\n",
    "\n",
    "\t.cache/\n",
    "\t__pycache__/\n",
    "\tbioinfo_dicts.py\n",
    "\tbioinfo_dicts.pyc\n",
    "\tseq_features.py\n",
    "\tseq_features.pyc\n",
    "\ttest_seq_features.py\n",
    "\n",
    "nothing added to commit but untracked files present (use \"git add\" to track)\n",
    "```\n",
    "\n",
    "That is where `.gitignore` comes into play. Let's not clutter our repo with all these other files that will be generated anyways.\n",
    "\n",
    "Make a new file called `.gitignore` (the dot here is important) and put the following in it:\n",
    "\n",
    "```sh\n",
    ".cache/\n",
    "__pycache__/ # There are two '_' in the beginning and at the end of this name\n",
    "*.pyc\n",
    "```\n",
    "\n",
    "This tells git to ignore files with extension `.pyc` and the directory `__pycache__`.\n",
    "\n",
    "```\n",
    "$ git status\n",
    "On branch master\n",
    "Your branch is up-to-date with 'origin/master'.\n",
    "Untracked files:\n",
    "  (use \"git add <file>...\" to include in what will be committed)\n",
    "\n",
    "\t.gitignore\n",
    "\tbioinfo_dicts.py\n",
    "\tseq_features.py\n",
    "\ttest_seq_features.py\n",
    "\n",
    "nothing added to commit but untracked files present (use \"git add\" to track)\n",
    "```\n",
    "\n",
    "Alright! Much cleaner! let's stage, commit and push:\n",
    "\n",
    "```\n",
    "$ git add .\n",
    "$ git commit -m \"Initial commit of protfeature\"\n",
    "$ git push origin master\n",
    "```\n",
    "\n",
    "Take a look at the page of your repository on github and these files should be there already. Now we are ready for CI."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Travis CI interface.\n",
    "\n",
    "Go to [travis CI webpage](http://travis-ci.org) and you should be seen something like this:\n",
    "\n",
    "<img src='l35_empty_travis.png' width=\"300px\">\n",
    "\n",
    "click on the `+` sign left of `My repositories`. That will take you to a page listing your repositories on GitHub (if you don't see anything, click in `Sync account`).\n",
    "\n",
    "<img src='l35_list_of_repos_travis.png' width=\"800px\">\n",
    "\n",
    "Toggle the CI for the repository we are interested in.\n",
    "\n",
    "<img src='l35_turn_it_on_travis.png' width=\"400px\">\n",
    "\n",
    "Go back to the home page on Travis CI and your repository should be listed there now but nothing happens. That is because we haven't told what Travis should do. Travis CI reads a `.travis.yml` file in the main direction with instructions on what it should do with your code. (The suffix `.yml` indicates that this is a YAML, pronounced \"yammal\" file, an acroynm for \"Yet Another Markup Language.\") Let's make the `.travis.yml` file:\n",
    "\n",
    "```yml\n",
    "language: python\n",
    "python:\n",
    "  - \"3.5\"\n",
    "  - \"3.6\"\n",
    "  \n",
    "# command to install dependencies\n",
    "\n",
    "# command to run tests\n",
    "script: py.test\n",
    "```\n",
    "\n",
    "This is probably the simplest `.travis.yml` possible. It tells which language our software is written oi, which versions should be able to run it and the testing command.\n",
    "\n",
    "Because we are not using any dependencies, there is nothing in the section `command to install dependencies` but if you have `import numpy` in your code, for example, you should include a line there saying something like: `pip install numpy`. If you have many dependencies, you should probably think about writing a `setup.py` file and use `pip` to help you keep track of all of them. This is a bit beyond the scope here, but if you are interested check out [pip freeze](https://pip.pypa.io/en/stable/reference/pip_freeze/).\n",
    "\n",
    "Now that you have added `.travis.yml`, `add`, `commit`, and `push`. Now, go to the Travis page and you should see something like this:\n",
    "\n",
    "<img src='l35_travis_working.png'>\n",
    "\n",
    "This is Travis cloning your repository and running your tests for you. If everything passes you should see something like this:\n",
    "\n",
    "<img src='l35_travis_passing.png'>\n",
    "\n",
    "and if it fails:\n",
    "\n",
    "<img src='l35_travis_failing.png'>\n",
    "\n",
    "You can also let people know if the current version of our program is passing the tests or not. For that, click on the badge `build` on the Travis page, select `Markdown`, copy the text and paste it on the README.md page on your repository.\n",
    "\n",
    "Next time you render the README.md page, it should look like this:\n",
    "\n",
    "<img src='l35_travis_badge.png' width=\"600px\">\n",
    "\n",
    "There, now your program is in continuous integration and everytime you receive a pull request, GitHub will let you know if their implementation passes all the tests."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Where do we go from here?\n",
    "\n",
    "There are tons of details about `pytest` and `travis-ci` that will address most issues you will encounter while working on your program. Both of them have very good resources on their page and are well documented.\n",
    "\n",
    "This lesson was heavily based on [Katy Huff's Software Carpentry Tutorial](http://katyhuff.github.io/python-testing/) and our own experience. Check it out for more details on TDD and CI.\n",
    "\n",
    "Finally, the next real step is for you to learn how to package your program and publish it (possibly on [pypi.python.org](pypi.python.org), or just hosted on GitHub). An interesting shortcut for that is to use the [Cookiecutter PyPackage](https://github.com/audreyr/cookiecutter-pypackage).\n",
    "\n",
    "Have fun, and always test and share your code."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
